### PR TITLE
Changed how the payload was sent to the gateway

### PR DESCRIPTION
**Pull request by alejandrofloresm**

### CODE REVIEW

**Review by rachelkirk — CHANGES_REQUESTED** [review]
When I run the remote tests with this change, they all display the following error:
test_failed_authorize(RemoteMitTest): NoMethodError: undefined method `bytesize`

When submitting a PR to ActiveMerchant it is expected that the unit and remote test output is included in the commit message so the reviewer has an idea of how the change is effecting the existing tests. 

When I run the remote tests on master they all pass without issue. The No Method Errors will need to be fixed before we can approve the change.

**Comment by EsporaInfra** [comment]
The new changes have two purposes. The first obeys the first set of changes as, given an update in how the payloads are handled by the gateway, adjustments need to be made to the way we are sending information in order to produce successful transactions. The second change is related to the scrubbing of the transcript. After testing several transactions and talking to the representatives of the gateway, we decided to remove the decrypting of the responses. This is because not only the responses do not contain sensitive information, but also due to the format in which the transcription is created, it appears that the encrypted string containing the responses is split sometimes and thus the encryption process generates errors. Is this what you meant by an explanation in the design changes? Thanks in advance.

File: test/remote/gateways/remote_mit_test.rb
Comment (rachelkirk) [inline test/remote/gateways/remote_mit_test.rb:120]: Is there a reason why this was changed from purchase to authorize?

**Review by rachelkirk — CHANGES_REQUESTED** [review]
Tests are passing, but I have a few questions about some of the changes. Thank you!

File: lib/active_merchant/billing/gateways/mit.rb
Comment (EsporaInfra) [reply lib/active_merchant/billing/gateways/mit.rb:185]: Of course. This code has been removed from the scrubbing function due to an issue with that segment of the response that presents issues when decrypted. Upon consulting it with the Gateway, we were told that particular part of the response holds no sensitive information and neither is useful for the purposes of debugging errors in transactions. It is with those considerations that we removed that specific segment of the response in the scrubbing, while the rest are left functional and any sensitive information removed after decryption.

File: test/remote/gateways/remote_mit_test.rb
Comment (EsporaInfra) [reply test/remote/gateways/remote_mit_test.rb:120]: This one is a mistake on our part and we are returning it to the previous version.

File: Gemfile
Comment (EsporaInfra) [reply Gemfile:13]: We have had an issue in which none of the tests can run unless this gem is added to the gemfile. It is the following error:

`LoadError: cannot load such file -- rexml/document`

It doesn;t appear to be related to any code of ours but on the file active_merchant.rb I can add the stack trace if needed.

Comment (rachelkirk) [reply Gemfile:13]: I don't get an error when I comment this gem out of the gemfile and run your tests. Also, as you mention it's already listed in required in the `active_merchant.rb` base file. Can you do me a favor and squash your commits to a single commit and then pull and rebase upstream/master? That's one of the Active Merchant conventions to keep the commit history clean. Several of the CI checks are failing and it might be related to an update that you haven't pulled in.

**Review by rachelkirk — CHANGES_REQUESTED** [review]
I don't think adding the rexml gem is necessary, I can run tests without the gem without a problem. You may just need to update your branch. Please squash to 1 commit to keep commit history clean and then pull and rebase upstream/master to see if that will get the CI checks passing.

**Review by rachelkirk — CHANGES_REQUESTED** [review]
Hi, It looks like you haven't updated the master branch in your fork and that's maybe what you rebased against instead of the main version of [ActiveMerchant](https://github.com/activemerchant/active_merchant). You should please do the following:
1. Add the main version of AM as a remote if they don’t have it:  `git remote add upstream git@github.com:activemerchant/active_merchant.git`
2. check out their local master branch:  `git checkout master`
3. pull master from the newly added remote: `git pull upstream master`
4. check the topic branch back out: `git checkout mit-change-payload`
5. rebase from the freshly updated master branch `git rebase master`
6. Force push your branch to github `git push force origin mit-change-payload`

Please let me know if you have trouble with this and I'll try to answer your questions. Thanks!

**Comment by EsporaInfra** [comment]
Hello. I've ran the commands as you suggested and they all ran without issues. Is everything in order now?

Thanks.

**Review by rachelkirk — CHANGES_REQUESTED** [review]
Thanks for making those changes. Things look good, but when I ran the remote tests I noticed there's an error in `test_transcript_scrubbing`. That will need to be addressed before I can approve and merge.

**Comment by rachelkirk** [comment]
> > Thanks for making those changes. Things look good, but when I ran the remote tests I noticed there's an error in `test_transcript_scrubbing`. That will need to be addressed before I can approve and merge.
> 
> Hello! We have run multiple test with no issue. Would it be possible for you to share with us the error you received in the scrubbing function? Were there any changes affecting the test's payload? Thank you.

I'm not sure what the issue was yesterday. The test is passing today. Here is some of the error output from my terminal when I ran the test yesterday:
`Error: test_transcript_scrubbing(RemoteMitTest): ArgumentError: iv must be 16 bytes
/Users/rachelkirk/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:50:in `iv='
/Users/rachelkirk/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:50:in `decrypt'
/Users/rachelkirk/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:190:in `block in scrub'
/Users/rachelkirk/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:186:in `loop'
/Users/rachelkirk/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:186:in `scrub'
test/remote/gateways/remote_mit_test.rb:123:in `test_transcript_scrubbing'
     120:       @gateway.purchase(@amount, @credit_card, @options_success)
     121:     end
     122: 
  => 123:     clean_transcript = @gateway.scrub(transcript)
     124:     assert_scrubbed(@credit_card.number, clean_transcript)
     125:     assert_scrubbed(@credit_card.verification_value, clean_transcript)
     126:     assert_scrubbed(@gateway.options[:api_key], clean_transcript)`
     
     Tests are passing now so I will approve and merge.

**Review by rachelkirk — APPROVED** [review]
Looks good and tests are passing

**Comment by rachelkirk** [comment]
I had some problems with rebasing the branch, it was out of date. I was able to get it fixed but will need to wait until tomorrow to merge it in since it's past our deploy window.

**Comment by rachelkirk** [comment]
Hey @alejandrofloresm, I just wanted to let you know that I've merged this PR into master.

File: test/unit/gateways/mit_test.rb
Comment (EsporaInfra) [reply test/unit/gateways/mit_test.rb:233]: We left that portion of the response encrypted due to an issue in which the encrypted responses from the gateway were altered (split into multiple sub strings) by the transcription and were left unable to be decrypted. This happens very rarely but it is the reason some of your remote tests had issues. Is there a way we can identify under what circumstances the generated transcription splits a string into substrings?

An example of what we saw would be like this

A response string would typically look like this:

-> "BnuAgMOx9USBreICk027VY2ZqJA7xQcRT9Ytz1WpabDnqIglj43J/I03pKLtDlFrerKIAzhW1YCroDOS7mvtA5YnWezLstoOK0LbIcYqLzj1dCFW2zLb9ssTCxJb7ZmEQdzQdl8pyY4qC0QQ0JrOrsSA9QfX1XhkdcSVnttQV1cEooL8/6EsVFCb6yVIMhVnGL6GRCc2J+rPigHsljLWRovgRKqFIURJjNWbbbbbDRPG2hCNKsabM/lE2DFtKLMs4J5iwY9HiRbrAMG6BaGNiQ=="

But the generated transcription we receive would look like this:

-> "BnuAgMOx9USBreICk027VY2ZqJA7xQcRT9Ytz1WpabDnqIglj43J/I03pKLtDlFrerKIAzhW1YCroDOS7mvtA5YnWezLstoOK0LbIcYqLzj1dCFW2zLb9ssTCxJb7ZmEQdzQdl8pyY4qC0QQ0JrOrsSA9QfX"
-> "1XhkdcSVnttQV1cEooL8/6EsVFCb6yVIMhVnGL6GRCc2J+rPigHsljLWRovgRKqFIURJjNWbbbbbDRPG2hCNKsabM/lE2DFtKLMs4J5iwY9HiRbrAMG6BaGNiQ=="

We don't know why sometimes we received a split encrypted answer and when we do not, even if the transaction is identical. If we decrypt the responses again this specific portion of the transcription would generate issues in the scrubbing function. We have passed this to the gateway in case it is an issue of theirs, but I would like to know if there's a case in which the transcription generation would split a string like that.

Comment (rachelkirk) [reply test/unit/gateways/mit_test.rb:233]: I just sent this message to Gabriela in an email but in case it doesn't get forwarded quickly:

We looked up the errors in the stack trace from the transaction you linked in a support thread:
Full error: "full_error_with_trace"=>"#<OpenSSL::Cipher::CipherError: wrong final block length>\n

The input byte length is not actually a multiple of 16 like the cipher expects. This is supported by running the remote test test_transcript_scrubbing and the test fails with ArgumentError: iv must be 16 bytes. You can run these remote tests and see this. 
Stacktrace:
ArgumentError: iv must be 16 bytes /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:50:in `iv=' /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:50:in `decrypt' /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:190:in `block in scrub' /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:186:in `loop' /Users/bsmith/dev/active_merchant/lib/active_merchant/billing/gateways/mit.rb:186:in `scrub' /Users/bsmith/dev/active_merchant/test/remote/gateways/remote_mit_test.rb:123:in `test_transcript_scrubbing'


The response splitting is not something that happens in ActiveMerchant, this is happening from the gateway before it gets to ActiveMerchant. A suggestion is to inspect where the encrypted response is concatenated and converted to base64 (here: https://github.com/activemerchant/active_merchant/blob/00f1189edab256b7a7d898368c0eeece2ddcbb1b/lib/active_merchant/billing/gateways/mit.rb#L75). 
 
After speaking with a colleague who is more versed in security and ciphers, they gave the following suggestions:

Possible solutions:
Adjust padding using the PKCS7 padding scheme to make up for byte lengths that are lesser than 16, with something like: engine.padding = 1 # Sets PKCS7 padding (src: [So What Is PKCS#7?](https://medium.com/asecuritysite-when-bob-met-alice/so-what-is-pkcs-7-daf8f4423fd1) )
Update the engine from CBC to prevent the need for padding in the first place (src: [AES-256, CBC, plaintext length is multiple of block size: Does PKCS #7 padding weaken the encryption?](https://crypto.stackexchange.com/a/82002) , [Incorrect AES Implementation Leaves System Vulnerable](https://www.winmill.com/blog/2022/08/05/incorrect-aes-implementation-leaves-system-vulnerable/) ).
:note: The engine from above is defined as engine = OpenSSL::Cipher::AES128.new(:CBC). I also wonder if they should graduate from AES-128 to AES-256.


I will be out of office next week but my teammates will be keeping an eye on this PR.

Comment (rachelkirk) [reply test/unit/gateways/mit_test.rb:233]: Please see the comments above.

**Review by rachelkirk — CHANGES_REQUESTED** [review]
Please see the comments and suggestions left a few weeks ago. Was this helpful?

File: test/unit/gateways/mit_test.rb
Comment (alejandrofloresm) [reply test/unit/gateways/mit_test.rb:233]: Hi @rachelkirk I think that we had found what it is the problem and we are looking for ways to solve it.

When the tests are ran the transcript is being stored on the transcript variable:

https://github.com/activemerchant/active_merchant/blob/92046f1680094d1b70557040d1c1edb193f7948d/test/remote/gateways/remote_mit_test.rb#L119

```
def test_transcript_scrubbing
    transcript = capture_transcript(@gateway) do
      @gateway.purchase(@amount, @credit_card, @options_success)
    end

    clean_transcript = @gateway.scrub(transcript)
    assert_scrubbed(@credit_card.number, clean_transcript)
    assert_scrubbed(@credit_card.verification_value, clean_transcript)
    assert_scrubbed(@gateway.options[:api_key], clean_transcript)
    assert_scrubbed(@gateway.options[:key_session], clean_transcript)
end
```

The method called `capture_transcript`, captures all the communication between Spreedly and MIT and stores it on the variable transcript (keep in mind that the method capture_transcript is not part of MIT integration)

Then on `@gateway.scrub` the variable `transcript` is passed. 

When reading the content of the variable `transcript` there are requests that sometimes are serialized on the `transcript` as with the full body of the response and sometimes the body of the response is being "broken". For example:

*Full*
```
-> "\r\n"
reading 344 bytes...
-> "VW0iCiVNSsdkxn0rBnESgK8lLiLnGxH/csfK4pnDVssQ1+QEs/XHWJMn9IMxnJehm1rn2UnOpq8Q4CAWrh9/xcVYmjc0N4BTrAISGcLHXQHskNth4cf3woREZ2lD2pOok2gmCbC/aEhhsHqO21G5NsZj09+5p7HiCTutiFnzFebkrBqJEV+g1Vm/oJHqMweU7wewFY6w4vHfKnSCBbTXfHygvbT98gugfm2/H6Y+d9EEOsu0nJb9eh9y9PC4tweoFA3vVe0spl0WIg14vexITkVN1HZIA5onw5SZtqW76JktGfRd7+K7L1oBjhWMjdmw2zYOpW4rEnQoAuzG90/sBw=="
read 344 bytes
Conn close
```


*Broken*
```
reading 344 bytes...
-> "lPRvJkNr1bg1"
-> "kHdqguEDHKCws4iDEEtLGaOz2/FZe7SOD4RliKkJzz61Lye80ts5ro13PKke+gEuHcR3hAu66m/bJRQCggmHFt3nIPyVrj9elZIXwaAcA1aSoijwTSzNwjjVQg+RQL9gYBweb86QvPUkmsKQFpyAP9lGQ1TTJLVKYGXqq+7cLiAaxdhehlx1N2Knckg7BDM8gP0j9ExsSXopaiBQ9ZZTzkIv2UURoFc+4AMMco/cFFXf9RyHkzER21EtxIs600ENcwhgsJ8GRjAhx6nNz+c5R5nCXK0xcsR/+KDpIHZUQEulrrb0VeLiJElsvhc++Uk5EHZp4Yq9mQ=="
read 344 bytes
```

Notice that on the "Broken" it shows two `->`.

So this is the reason why when the tests are running `clean_transcript = @gateway.scrub(transcript)` it fails. Because when integration is reading the response from the MIT server from the `transcript` it fails  since its not the response from the MIT Server: https://github.com/activemerchant/active_merchant/blob/92046f1680094d1b70557040d1c1edb193f7948d/lib/active_merchant/billing/gateways/mit.rb#L173

Do you have any recommendation regarding on how to avoid that the response body is added incorrectly into the transcript?

Thanks!

Comment (rachelkirk) [reply test/unit/gateways/mit_test.rb]: I'm still looking into this, but I have not seen any instances of ActiveMerchant splitting a response like this before. This may have something to do with the way the encrypt and decrypt methods are written. There are only a few gateways out of 150+ in ActiveMerchant that have encrypt and decrypt methods defined like this. The helper method that you wrote is returning the auth and capture response simultaneously for a purchase since it's a `MultiResponse.run` of auth and capture for purchase. I'm doing more testing locally to see if that will be an issue.

File: test/unit/gateways/mit_test.rb
Comment (rachelkirk) [inline test/unit/gateways/mit_test.rb:239]: Why is your helper method returning the capture response after the authorize POST? It's returning each response twice. It seems like only the decrypted authorize response should be here.

File: test/unit/gateways/mit_test.rb
Comment (rachelkirk) [inline test/unit/gateways/mit_test.rb:264]: And it seems like only the decrypted capture response should be here.

**Review by rachelkirk — CHANGES_REQUESTED** [review]
I'm still looking into this and doing some testing locally but had a few questions

**Comment by alejandrofloresm** [comment]
> @alejandrofloresm Can you please link to the places where you were demonstrating that the split is happening during our call? I have not investigated these issues within the ruby net http library as part of my work for this ticket since it is primarily to review your changes, but that would be helpful for when I bring this to our security team.

Sure, you can find it in here:

Reads the response:
https://github.com/ruby/ruby/blob/ruby_2_7/lib/net/protocol.rb#L151-L169

The response is being written in chunks:
https://github.com/ruby/ruby/blob/ruby_2_7/lib/net/protocol.rb#L229

File: test/unit/gateways/mit_test.rb
Comment (rachelkirk) [reply test/unit/gateways/mit_test.rb:239]: I don't know if this is a blocking change but I'm still wondering why this is happening, seems unusual.

**Review by rachelkirk — CHANGES_REQUESTED** [review]
Our security team has reviewed and approved your solution for joining the transcript when split. I left a comment about the `rexml` gem. It was added as a dependency a few months ago so you should be able to remove your addition to the Gemfile. Please squash your commits down to 1 and rebase against upstream/master and then I'll be able to merge this in. Thank you!

**Comment by rachelkirk** [comment]
> Hi, It looks like you haven't updated the master branch in your fork and that's maybe what you rebased against instead of the main version of [ActiveMerchant](https://github.com/activemerchant/active_merchant). You should please do the following:
> 
> 1. Add the main version of AM as a remote if they don’t have it:  `git remote add upstream git@github.com:activemerchant/active_merchant.git`
> 2. check out their local master branch:  `git checkout master`
> 3. pull master from the newly added remote: `git pull upstream master`
> 4. check the topic branch back out: `git checkout mit-change-payload`
> 5. rebase from the freshly updated master branch `git rebase master`
> 6. Force push your branch to github `git push force origin mit-change-payload`
> 
> Please let me know if you have trouble with this and I'll try to answer your questions. Thanks!

Hey just checking in, it looks like you may need to try the steps I suggested above to get your branch in a good state. Our change cut-off is 3:00 EST time and this will require two merges so I'm not sure we'll make the cutoff for today. Please let me know if I can help.

**Review by rachelkirk — APPROVED** [review]
I'm approving this, but waiting for approval to do both merges. I had to pull the branch down and then run tests and by the time I finished reviewing the change cutoff had passed. If it can't be merged today I'll ask someone from my team merge it next week. (I will be out of office)

**Pull request closed without merging**
